lib/gtfs.ex

```diff
@@ -0,0 +1,165 @@
+defmodule Gtfs do
+  use GenServer
+  require Logger
+
+  @type t :: %__MODULE__{
+          routes: [Gtfs.Route.t()],
+          stop_times: [Gtfs.StopTime.t()],
+          stops: [Gtfs.Stop.t()],
+          trips: [Gtfs.Trip.t()]
+        }
+
+  @enforce_keys [
+    :routes,
+    :stop_times,
+    :stops,
+    :trips
+  ]
+
+  defstruct [
+    :routes,
+    :stop_times,
+    :stops,
+    :trips
+  ]
+
+  @type state :: t
+
+  @type query_opts :: %{
+          optional(:server) => GenServer.server()
+        }
+
+  @type files :: %{optional(String.t()) => binary()}
+
+  @type files_source :: {:url, String.t()} | {:mocked_files, mocked_files()}
+
+  @typedoc """
+  For mocking tests
+  E.g.
+  %{
+    "stops.txt" => [
+      "stop_id,stop_name",
+      "place-sstat,South Station",
+    ]
+  ]}
+  """
+  @type mocked_files :: %{optional(String.t()) => [binary()]}
+
+  # Client functions
+
+  @spec start_link(String.t()) :: GenServer.on_start()
+  def start_link(url) do
+    GenServer.start_link(__MODULE__, {:url, url}, name: __MODULE__)
+  end
+
+  @spec start_mocked(mocked_files) :: GenServer.on_start()
+  def start_mocked(mocked_files) do
+    GenServer.start_link(__MODULE__, {:mocked_files, mocked_files})
+  end
+
+  @doc """
+  Returns all of GTFS, with each file as an item in a map
+  """
+  @spec gtfs(query_opts()) :: t()
+  def gtfs(opts) do
+    server = opts[:server] || __MODULE__
+    GenServer.call(server, :state)
+  end
+
+  # GenServer callbacks
+
+  @impl true
+  def init(files_source) do
+    case fetch_files(files_source) do
+      {:err, err} ->
+        {:stop, err}
+
+      files ->
+        state = parse_files(files)
+        Logger.info(fn -> "Successfully loaded gtfs" end)
+        {:ok, state}
+    end
+  end
+
+  @impl true
+  def handle_call(:state, _from, state) do
+    {:reply, state, state}
+  end
```

Comment:
You probably want calls to return particular parts of the state, rather than the whole thing. Let the GenServer encapsulate the state, and have the clients request particular pieces they're looking for. In this case, something like `handle_call(:routes, _from, state), do: {:reply, state.routes, state}`

Reply:
I suppose, but even then you probably want to provide helper functions to `Gtfs.Query`. Otherwise, now you've got two modules which need to know about the shape of the state here, rather than it being encapsulated by `Gtfs`. It also sets you up better when there's more data here, as calls won't need to copy the entire state to another process when a request comes in.

Reply:
That's why I clearly defined the shape of the data in `Gtfs.t()`. I think that sharing type between `Gtfs.Query` and `Gtfs` is less trouble than trying to keep query logic organized in `gtfs.ex`.

Will returning the state require copying all the data, or would it be a reference to shared memory?

Reply:
It's copied. From http://erlang.org/doc/efficiency_guide/processes.html#process-messages

> All data in messages between Erlang processes is copied, except for refc binaries on the same Erlang node.

If you want to keep is this way, I'd flip how the state/querying logic is split.
* Gtfs.Query: responsible for managing the struct, and returning results based on that struct (no process)
* Gtfs: responsible for fetching/keeping the Gtfs.Query struct, and calls functions in Gtfs.Query to return results.

For example:
```elixir
defmodule Gtfs do
  def routes(server \\ __MODULE__) do
    GenServer.call(server, :routes)
  end

  def handle_continue(:fetch, state) do
    with {:ok, %{body: body}} <- HTTPoison.get(url()) do
      new_state = Gtfs.Query.from_zip_file(body)
      {:noreply, new_state}
    else
      _ -> {:stop, :normal, state}
  end

  def handle_call(:routes, _from, state) do
    routes = Gtfs.Query.routes(state)
    {:reply, routes, state}
  end
end

defmodule Gtfs.Query do
  defstruct [:routes, :trips, ...]

  def from_zip_file(zip_file) do
    %__MODULE__{routes: ...}
  end

  def routes(%__MODULE__{routes: routes}), do: routes
end
```

Reply:
Discussion continued in slack: https://mbtace.slack.com/archives/CESHREAS1/p1554216526039500

Reply:
Addressed by moving `Gtfs.Query` into `Gtfs`. There now needs to be two functions per query, and the queries are mixed in the same file as the genserver. That's fine for now. If it becomes too unwieldy when we have many more queries, we can look at splitting them again.